util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,412 @@
+		MaxStatementsNum:   config.MaxStatementsNum,
+		collectInterval:    config.CollectInterval,
+		collectTimeout:     collectTimeout,
+		topSQLMap:          make(map[string]*topSQLDataPoints),
+		normalizedSQLMap:   make(map[string]string),
+		normalizedPlanMap:  make(map[string]string),
+		collectCPUTimeChan: make(chan *topSQLCPUTimeInput, 1),
+		planRegisterChan:   make(chan *planRegisterJob, 10),
+		agentGRPCAddress:   config.AgentGRPCAddress,
+		instanceID:         config.InstanceID,
+		quit:               make(chan struct{}),
+	}
+
+	go tsr.collectWorker()
+
+	go tsr.registerPlanWorker()
+
+	go tsr.sendToAgentWorker()
+
+	return tsr
+}
+
+// Collect will drop the records when the collect channel is full
+func (tsr *RemoteTopSQLReporter) Collect(timestamp uint64, records []tracecpu.TopSQLCPUTimeRecord) {
+	select {
+	case tsr.collectCPUTimeChan <- &topSQLCPUTimeInput{
+		timestamp: timestamp,
+		records:   records,
+	}:
+	default:
+	}
+}
+
+func (tsr *RemoteTopSQLReporter) collectWorker() {
+	for {
+		input := <-tsr.collectCPUTimeChan
+		tsr.collect(input.timestamp, input.records)
+	}
+}
+
+// collect uses a hashmap to store records in every second, and evict when necessary.
+// This function can be run in parallel with snapshot, so we should protect the map operations with a mutex.
+func (tsr *RemoteTopSQLReporter) collect(timestamp uint64, records []tracecpu.TopSQLCPUTimeRecord) {
+	for _, record := range records {
+		encodedKey := encodeCacheKey(record.SQLDigest, record.PlanDigest)
+		entry, exist := tsr.topSQLMap[string(encodedKey)]
+		if !exist {
+			entry = &topSQLDataPoints{
+				SQLDigest:     record.SQLDigest,
+				PlanDigest:    record.PlanDigest,
+				CPUTimeMsList: []uint32{record.CPUTimeMs},
+				TimestampList: []uint64{timestamp},
+			}
+			tsr.topSQLMap[string(encodedKey)] = entry
+		} else {
+			entry.CPUTimeMsList = append(entry.CPUTimeMsList, record.CPUTimeMs)
+			entry.TimestampList = append(entry.TimestampList, timestamp)
+		}
+		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
+	}
+
+	if len(tsr.topSQLMap) <= tsr.MaxStatementsNum {
+		return
+	}
+
+	// find the max CPUTimeMsTotal that should be evicted
+	digestCPUTimeList := make([]cpuTimeSort, len(tsr.topSQLMap))
+	{
+		i := 0
+		for key, value := range tsr.topSQLMap {
+			data := cpuTimeSort{
+				Key:            key,
+				SQLDigest:      value.SQLDigest,
+				PlanDigest:     value.PlanDigest,
+				CPUTimeMsTotal: value.CPUTimeMsTotal,
+			}
+			digestCPUTimeList[i] = data
+			i++
+		}
+	}
+
+	// QuickSelect will only return error when the second parameter is out of range
+	if err := quickselect.QuickSelect(cpuTimeSortSlice(digestCPUTimeList), tsr.MaxStatementsNum); err != nil {
+		//	skip eviction
+		return
+
+	}
+
+	// TODO: we can change to periodical eviction (every minute) to relax the CPU pressure
+	shouldEvictList := digestCPUTimeList[tsr.MaxStatementsNum:]
+	for _, evict := range shouldEvictList {
+		delete(tsr.topSQLMap, evict.Key)
```

Comment:
This need to be protected behind a lock.

Reply:
fixed